### ISSUE TITLE
Helioprojective coordinate overlay shows wrong units

### ISSUE THREAD

**Issue by svank** [sunpy/sunpy#7653]
### Describe the bug

Not sure if this is an astropy or sunpy bug (or user error).

If I make a plot with `WCSAxes` and do a coordinate overlay of helioprojective coordinates, it looks like the helioprojective grid is in degrees, but it's marked as arcseconds.

I'm not sure how to make a WCS with a builtin frame and get the helioprojective origin to be the same, so here's an example with a custom frame that's just helioprojective, but rotated. (I encountered this when using my own custom frame that transforms to helioprojective).
![image](https://github.com/sunpy/sunpy/assets/23462789/a5dd2820-4c82-4933-bd30-0f851564293a)

Note that the custom frame ticks are every five degrees, and the helioprojective overlay ticks have the same spacing, but they're labeled as every 5 arcsec. (And since the custom frame is just rotated helioproejctive coords, the two frames' degrees should be the same size on the plot.)

Swapping which frame is the overlay gives a plot that looks right:
![image](https://github.com/sunpy/sunpy/assets/23462789/543b9005-bc31-451f-8833-106906bcaf25)

I wonder if this has to do with helioprojective using arcsec as its default unit.

### To Reproduce

```python
import astropy.coordinates
import astropy.units as u
from astropy.wcs import WCS
import matplotlib.pyplot as plt
import numpy as np
import sunpy.coordinates

#
# Create a frame that transforms to helioprojective
#

class TiltedFrame(astropy.coordinates.BaseCoordinateFrame):
    """
    This is a Helioprojective frame, but rotated 20 degrees around the Sun--observer axis
    """
    default_representation = astropy.coordinates.representation.SphericalRepresentation
    obstime = astropy.coordinates.TimeAttribute()
    observer = sunpy.coordinates.frameattributes.ObserverCoordinateAttribute(
        sunpy.coordinates.HeliographicStonyhurst)


@astropy.coordinates.frame_transform_graph.transform(
    astropy.coordinates.DynamicMatrixTransform,
    sunpy.coordinates.Helioprojective,
    TiltedFrame)
def hpc_to_tilted(HPcoord, tilted_frame):
    angle = 20*u.deg
    matrix = np.array([[1, 0, 0],
                       [0, np.cos(angle), -np.sin(angle)],
                       [0, np.sin(angle),  np.cos(angle)]])
    return matrix


@astropy.coordinates.frame_transform_graph.transform(
    astropy.coordinates.DynamicMatrixTransform,
    TiltedFrame,
    sunpy.coordinates.Helioprojective)
def tilted_to_hpc(tilted_coord, HPframe):
    angle = -20*u.deg
    matrix = np.array([[1, 0, 0],
                       [0, np.cos(angle), -np.sin(angle)],
                       [0, np.sin(angle),  np.cos(angle)]])
    return matrix

#
# Register the frame
#

def tilted_wcs_frame_mapping(wcs):
    ctypes = {c[:4] for c in wcs.wcs.ctype}
    if ({'TLLN', 'TLLT'} <= ctypes):
        dateobs = wcs.wcs.dateavg or wcs.wcs.dateobs or None
        observer = sunpy.coordinates.HeliographicStonyhurst(
            wcs.wcs.aux.hgln_obs * u.deg,
            wcs.wcs.aux.hglt_obs * u.deg,
            wcs.wcs.aux.dsun_obs * u.m)
        return TiltedFrame(observer=observer, obstime=dateobs)
    return None


def tilted_frame_wcs_mapping(frame, projection='TAN'):
    if isinstance(frame, TiltedFrame):
        wcs = astropy.wcs.WCS(naxis=2)
        wcs.wcs.dateobs = frame.obstime.utc.isot
        wcs.wcs.aux.hgln_obs = frame.lon.to_value(u.deg)
        wcs.wcs.aux.hglt_obs = frame.lat.to_value(u.deg)
        wcs.wcs.aux.dsun_obs = frame.radius.to_value(u.m)
        wcs.wcs.ctype = f'TLLN-{projection}', f'TLLT-{projection}'
        wcs.wcs.cunit = ['deg', 'deg']
        return wcs
    return None


astropy.wcs.utils.WCS_FRAME_MAPPINGS.append([tilted_wcs_frame_mapping])
astropy.wcs.utils.FRAME_WCS_MAPPINGS.append([tilted_frame_wcs_mapping])
astropy.wcs.wcsapi.fitswcs.CTYPE_TO_UCD1_CUSTOM.append(
    {
        "TLLN": "custom:pos.tiltedframe.lon",
        "TLLT": "custom:pos.tiltedframe.lat",
    })


#
# Create the "main" WCS for plotting
#

wcs = WCS(naxis=2)
wcs.wcs.ctype = "TLLN-CAR", 'TLLT-CAR'
wcs.wcs.crpix = 100, 100
wcs.wcs.crval = 0, 0
wcs.wcs.cdelt = 0.1, 0.1
wcs.wcs.aux.hgln_obs = 0
wcs.wcs.aux.hglt_obs = 0
wcs.wcs.aux.dsun_obs = u.au.to(u.m)
wcs.wcs.dateobs = "2020-01-01T00:00:00"
wcs.wcs.dateavg = wcs.wcs.dateobs
wcs.fix()


ax = plt.subplot(111, projection=wcs)
plt.imshow(np.zeros((200, 200)))
ax.coords.grid(color='white')
ax.coords[0].set_axislabel("Custom frame X")
ax.coords[1].set_axislabel("Custom frame Y")

overlay = ax.get_coords_overlay('helioprojective')
overlay.grid(color='gold', ls='dotted')
overlay[0].set_axislabel("Helioprojective Tx")
overlay[1].set_axislabel("Helioprojective Ty")
```

### System Details

This occurs with the latest release of sunpy and astropy, and with current git for both.
```
==============================
sunpy Installation Information
==============================

General
#######
OS: Ubuntu (22.04, Linux 6.5.0-28-generic)
Arch: 64bit, (x86_64)
sunpy: 5.1.3
Installation path: /home/svankooten/.anaconda/envs/test-env/lib/python3.12/site-packages/sunpy-5.1.3.dist-info

Required Dependencies
#####################
astropy: 6.1.0
numpy: 1.26.4
packaging: 24.0
parfive: 2.1.0
pyerfa: 2.0.1.4

Optional Dependencies
#####################
asdf: 3.2.0
asdf-astropy: 0.6.1
beautifulsoup4: 4.12.3
cdflib: 0.0.0
dask: 2024.5.1
drms: 0.6.4
glymur: 0.13.2.post1
h5netcdf: 1.3.0
h5py: 3.11.0
lxml: 5.2.2
matplotlib: 3.8.4
mpl-animators: 1.1.1
pandas: 2.2.2
python-dateutil: 2.9.0
reproject: 0.13.1
scikit-image: 0.22.0
scipy: 1.13.1
sqlalchemy: 2.0.30
tqdm: 4.66.4
zeep: 4.2.1
```

### Installation method

conda

**Comment by svank** [comment]
It appears I can work around this like so, but I have no idea how reliable or robust this is

```python
ax = plt.subplot(111, projection=wcs)
plt.imshow(np.zeros((200, 200)))
ax.coords.grid(color='white')
ax.coords[0].set_axislabel("Custom frame X")
ax.coords[1].set_axislabel("Custom frame Y")

overlay = ax.get_coords_overlay('helioprojective')
overlay.grid(color='gold', ls='dotted')
overlay[0].set_axislabel("Helioprojective Tx")
overlay[1].set_axislabel("Helioprojective Ty")

# Workaround
lon, lat = overlay
# Override the unit
lon.coord_unit = u.deg
lat.coord_unit = u.deg
# Setting tick location is required, otherwise only the zero-degree ticks/lines show up
lat.set_ticks(np.arange(-90, 90, 5) * u.deg)
lon.set_ticks(np.arange(-180, 180, 5) * u.deg)
# Optional, otherwise, HP ticks are labeled as a correct number of arcsec, while still spaced 5 degrees
lat.set_major_formatter('dd')
lon.set_major_formatter('dd')
```
![image](https://github.com/sunpy/sunpy/assets/23462789/93bc344f-11d7-4e75-8055-be352612cd86)

**Comment by ayshih** [comment]
I found the bug in WCSAxes, so I'll put up a PR to Astropy soon.  If I run your example code, the output is now:
![download (13)](https://github.com/sunpy/sunpy/assets/991759/619db093-c325-4eb2-a3b7-bf74098c812e)
The units for the overlay can be changed to degrees by calling:
```python
overlay[0].set_format_unit(u.deg)
overlay[1].set_format_unit(u.deg)
```

**Comment by svank** [comment]
Wonderful, thanks!

**Comment by ayshih** [comment]
See astropy/astropy#16662

**Comment by ayshih** [comment]
Closing because the fix upstream was released in astropy 6.1.2
